File: temperature_converter.js
function temperatureConverter() {
    let count = 0;  // Enclosed variable

    return function convert(temp, toCelsius = true) {
        count++;  // Modifying enclosed variable
        return toCelsius
            ? [(temp - 32) * 5 / 9, count]  // Convert to Celsius
            : [(temp * 9 / 5) + 32, count]; // Convert to Fahrenheit
    };
}

const convertTemp = temperatureConverter();
console.log(`100 degrees F is ${convertTemp(100, true)[0]} C`)
console.log(`0 degrees C is ${convertTemp(0, false)[0]} F`)

File: Section_3/temperature_converter.rs
use std::rc::Rc;
use std::cell::RefCell;

struct TemperatureConverter {
    count: Rc<RefCell<u32>>, // Reference-counted memory with interior mutability
}

impl TemperatureConverter {
    fn new() -> Self {
        TemperatureConverter {
            count: Rc::new(RefCell::new(0)), // Initialize with RefCell
        }
    }

    fn convert(&self, temp: f64, to_celsius: bool) -> (f64, u32) {
        let mut count = self.count.borrow_mut(); // Borrow mutably
        *count += 1; // Increment count

        let result = if to_celsius {
            (temp - 32.0) * 5.0 / 9.0
        } else {
            (temp * 9.0 / 5.0) + 32.0
        };

        (result, *count)
    }
}

fn main() {
    let converter = TemperatureConverter::new();
    let (converted1, calls1) = converter.convert(100.0, true);
    println!("Converted: {:.2}, Calls: {}", converted1, calls1);

    let (converted2, calls2) = converter.convert(0.0, false);
    println!("Converted: {:.2}, Calls: {}", converted2, calls2);
}
